pkg/liqo-controller-manager/external-network/wireguard/wggatewayserver_controller.go

```diff
@@ -0,0 +1,370 @@
+		if err != nil {
+			return err
+		}
+	case corev1.ServiceTypeLoadBalancer:
+		endpointStatus = r.forgeEndpointStatusLoadBalancer(service)
+	default:
+		err := fmt.Errorf("service type %q not supported for WireGuard server Service %q", service.Spec.Type, svcNsName)
+		klog.Error(err)
+		// TODO: Should we empty the endpoint status and force requeue instead of returning error (misaligned spec and status)?
+		return err
+	}
+
+	// Requeue if no endpoint is found at the moment
+	if endpointStatus != nil && len(endpointStatus.Addresses) == 0 {
+		*requeue = true
+	}
+
+	// Do not update status if endpoint has not changed
+	if !reflect.DeepEqual(wgServer.Status.Endpoint, endpointStatus) {
+		wgServer.Status.Endpoint = endpointStatus
+		*needStatusUpdate = true
+	}
+
+	return nil
+}
+
+func (r *WgGatewayServerReconciler) getWgServerService(ctx context.Context, svcName types.NamespacedName) (*corev1.Service, error) {
+	var svc corev1.Service
+	err := r.Get(ctx, svcName, &svc)
+	switch {
+	case err != nil && !apierrors.IsNotFound(err):
+		klog.Errorf("Unable to get WireGuard server service %s: %v", svcName, err)
+		return nil, err
+	case apierrors.IsNotFound(err):
+		klog.Infof("WireGuard server service %s not found. Retrying later", svcName)
+		return nil, nil
+	default:
+		return &svc, nil
+	}
+}
+
+func (r *WgGatewayServerReconciler) forgeEndpointStatusNodePort(ctx context.Context, service *corev1.Service,
+	dep *appsv1.Deployment, requeue *bool) (*networkingv1alpha1.EndpointStatus, error) {
+	port := service.Spec.Ports[0].NodePort
+	protocol := &service.Spec.Ports[0].Protocol
+	var addresses []string
+
+	// Every node IP is a valid endpoint. For convenience, we get the IP of all nodes hosting replicas of the deployment
+	// (i.e., WireGuard gateway servers).
+	podsFromDepSelector := client.MatchingLabelsSelector{Selector: labels.SelectorFromSet(dep.Spec.Selector.MatchLabels)}
+	var podList corev1.PodList
+	if err := r.extNetPodsClient.List(ctx, &podList, client.InNamespace(dep.Namespace), podsFromDepSelector); err != nil {
+		klog.Errorf("Unable to list pods of deployment %s/%s: %v", dep.Namespace, dep.Name, err)
+		return nil, err
+	}
+	switch len(podList.Items) {
+	case 0:
+		// requeue later
+		klog.Infof("Pods of deployment %s/%s not found. Retrying later", dep.Namespace, dep.Name)
+		*requeue = true
+		return nil, nil
+	default:
+		// TODO: if using active-passive, it should get the IP of the active node
+		var nodes []corev1.Node
+		for i := range podList.Items {
+			pod := &podList.Items[i]
+			klog.Infof("%s %s", pod.Name, pod.Spec.NodeName)
+			// get node hosting pod
+			var node corev1.Node
+			err := r.Get(ctx, types.NamespacedName{Name: pod.Spec.NodeName}, &node)
+			if err != nil && !apierrors.IsNotFound(err) {
+				klog.Errorf("Unable to get node %q: %v", pod.Spec.NodeName, err)
+				return nil, err
+			}
+			if !apierrors.IsNotFound(err) && utils.IsNodeReady(&node) { // TODO: what if the node is temporarily not ready?
```

Comment:
i think we should take a random node in that case... the alternative is to trigger a new reconciliation when it comes up, but i don't find any way doing that